src/vm/genericdict.cpp

```diff
@@ -109,183 +109,333 @@ DictionaryLayout::GetFirstDictionaryBucketSize(
+            *EnsureWritablePages(&(pDictLayout->m_slots[iSlot].m_signature)) = pResultSignature;
+            *EnsureWritablePages(&(pDictLayout->m_slots[iSlot].m_signatureSource)) = signatureSource;
+
+            pResult->signature = pDictLayout->m_slots[iSlot].m_signature;
+
+            _ASSERTE(FitsIn<WORD>(nFirstOffset + 1));
+            pResult->indirections = static_cast<WORD>(nFirstOffset + 1);
+            pResult->offsets[nFirstOffset] = slot * sizeof(DictionaryEntry);
+            *pSlotOut = slot;
+            return TRUE;
+        }
+
+        slot++;
     }
-} // DictionaryLayout::FindToken
 
+    *pSlotOut = pDictLayout->m_numSlots;
+    return FALSE;
+}
+
+#ifndef CROSSGEN_COMPILE
 /* static */
-BOOL 
-DictionaryLayout::FindToken(LoaderAllocator *               pAllocator, 
-                            DWORD                           numGenericArgs, 
-                            DictionaryLayout *              pDictLayout, 
-                            CORINFO_RUNTIME_LOOKUP *        pResult, 
-                            SigBuilder *                    pSigBuilder, 
-                            int                             nFirstOffset, 
-                            DictionaryEntrySignatureSource  signatureSource)
+DictionaryLayout* DictionaryLayout::ExpandDictionaryLayout(LoaderAllocator*                 pAllocator, 
+                                                           DictionaryLayout*                pCurrentDictLayout, 
+                                                           DWORD                            numGenericArgs, 
+                                                           SigBuilder*                      pSigBuilder, 
+                                                           BYTE*                            pSig, 
+                                                           int                              nFirstOffset, 
+                                                           DictionaryEntrySignatureSource   signatureSource, 
+                                                           CORINFO_RUNTIME_LOOKUP*          pResult,
+                                                           WORD*                            pSlotOut)
 {
-    WRAPPER_NO_CONTRACT;
+    CONTRACTL
+    {
+        STANDARD_VM_CHECK;
+        INJECT_FAULT(ThrowOutOfMemory(););
+        PRECONDITION(SystemDomain::SystemModule()->m_DictionaryCrst.OwnedByCurrentThread());
+        PRECONDITION(CheckPointer(pResult) && CheckPointer(pSlotOut));
+    }
+    CONTRACTL_END
+        
+    // There shouldn't be any empty slots remaining in the current dictionary.
+    _ASSERTE(pCurrentDictLayout->m_slots[pCurrentDictLayout->m_numSlots - 1].m_signature != NULL);
+
+    // Expand the dictionary layout to add more slots
+    _ASSERTE(FitsIn<WORD>(pCurrentDictLayout->m_numSlots + NUM_DICTIONARY_SLOTS));
+    DictionaryLayout* pNewDictionaryLayout = Allocate(pCurrentDictLayout->m_numSlots + NUM_DICTIONARY_SLOTS, pAllocator, NULL);
```

Comment:
`+ NUM_DICTIONARY_SLOTS` growth strategy will lead to quadratic memory consumption in pathological cases. For example, consider a pathological method that needs 100 different dictionary slots. This should do multiplication by a factor instead.

Reply:
Ok, I'll change this to a multiplication by a factor of 2, to double the size each time.